app/src/main/java/co/techmagic/hr/presentation/time_tracker/HrAppTimeTrackerPresenter.kt

```diff
@@ -1,24 +1,168 @@
+
+    private val cache: HashMap<String, MutableList<UserReportViewModel>> = HashMap(7)
+    private val holidays: HashMap<String, Holiday> = HashMap()
+    private val subscriptions: HashMap<String, Subscription> = HashMap(7)
+
+    var currentDate: Calendar = dateTimeProvider.now().dateOnly()
+    var selectedDate: Calendar = currentDate.copy()
+
+    override fun onViewCreated(isInitial: Boolean) {
+        super.onViewCreated(isInitial)
+        view?.init(selectedDate)
+    }
+
+    override fun onViewDestroyed() {
+        super.onViewDestroyed()
+        for (s in subscriptions.values) {
+            s.unsubscribe()
+        }
+        subscriptions.clear()
+    }
 
     override fun onWeekSelected(firstDayOfWeek: Calendar) {
-        // TODO: implement
+        val date = firstDayOfWeek.copy()
+        val selectedDayIndex = selectedDate.get(Calendar.DAY_OF_WEEK)
+        date.add(Calendar.DAY_OF_WEEK, if (selectedDayIndex == Calendar.SUNDAY) 6 else selectedDayIndex - Calendar.MONDAY ) //todo: refactor!
+        onDateSelected(date.copy())
+    }
+
+    override fun onDateSelected(year: Int, monthOfYear: Int, dayOfMonth: Int) {
+        val date = dateTimeProvider.now().dateOnly()
+        date.set(year, monthOfYear, dayOfMonth)
+        onDateSelected(date)
     }
 
     override fun onDateSelected(date: Calendar) {
-        // TODO: implement
+        if (selectedDate.isSameDate(date)) return
+
+        selectedDate = date.copy()
+        view?.notifyWeekDataChanged(date.firstDayOfWeekDate())
+        view?.selectWeek(date.firstDayOfWeekDate())
+        view?.selectDay(date.copy())
+
+        view?.showToolbarTitle(date.formatDate(TOOLBAR_DATE_FORMAT))
+    }
+
+    override fun onBindWeek(weekView: TimeTrackerWeekView, firstDayOfWeek: Calendar) {
+        weekView.setSelectedDay(selectedDate)
+        forWeek(firstDayOfWeek) { date ->
+            val key = key(date)
+            weekView.setTotalTime(date, totalDayMinutes(key), holidays[key])
+        }
     }
 
     override fun onBindDay(day: TimeTrackerDayView, date: Calendar) {
-        // TODO: implement logic to show reports for selected day
-        // day.showReports()
+        val dayReports = getCachedReports(date)
+        if (dayReports != null) {
+            if (dayReports.isEmpty()) {
+                day.showEmptyMessage(quotesManager.getRandomFormattedQuote())
+            } else {
+                day.showReports(dayReports)
+            }
+        } else {
+            day.showEmptyMessage(quotesManager.getRandomFormattedQuote())
+
+            val user = SharedPreferencesUtil.readUser() // TODO: refactor - inject account manager instead!!!
+            val firstDayOfWeek = date.firstDayOfWeekDate()
+            val weekReportsKey = key(firstDayOfWeek)
+            subscriptions[weekReportsKey]?.unsubscribe()
+            subscriptions[weekReportsKey] = timeReportRepository.getDayReports(user.id, firstDayOfWeek)
```

Comment:
Is it OK to use repository in presenter without some Interactor or UseCase? 
If we skip Interactors and/or UseCases we will skip business layer.
And yes, I know that they usually looks like only wrappers :)

Reply:
I had no intention to skip interactors it was just faster to implement logic in this way. I'll leave a comment to refactor everything in a proper way.